### PR TITLE
Remove FactorMetaData

### LINKED ISSUE #1563: Remove FMd (should be replaced by CalcFactor)
~First CPT~ #1562 

This first and then #1562

### PR DESCRIPTION
**Pull request by Affie**
Outstanding: 
- [x] should `fullvariables` be filtered by hypo
- [x] `solvefor` weirdness in DERelative
- [x] update news for breaking changes
- [x] bump minor version

Also, CPT should be removed in the same release to avoid possible issues.

This will close #1563

### CODE REVIEW

File: src/entities/FactorOperationalMemory.jl [hunk 1]
```diff
@@ -42,6 +39,12 @@ struct CalcFactor{T <: AbstractFactor, M, P <: Union{<:Tuple,Nothing,AbstractVec
   _allowThreads::Bool
   """ user cache of arbitrary type, overload the [`preambleCache`](@ref) function. NOT YET THREADSAFE """
   cache::C
+
+  ## TODO Consolidation WIP with FactorMetadata
+  # full list of variables connected to the factor
+  fullvariables::Vector{DFGVariable}
```
Comment (dehann) [inline src/entities/FactorOperationalMemory.jl:45]: ah dynamic types.  Let's do Vector just to get rid of FMd and the come back for `{T<:Tuple}`?

File: src/entities/FactorOperationalMemory.jl [hunk 2]
```diff
@@ -181,6 +153,8 @@ mutable struct CommonConvWrapper{ T<:AbstractFactor,
   _gradients::G
   # type used for cache
   dummyCache::CT
+  #
+  fullvariables::Vector{DFGVariable}
```
Comment (dehann) [inline src/entities/FactorOperationalMemory.jl:157]: yeah, cleanup of CCW after FMd and CPT are removed.  That should start to help performance, allocations, dynamic dispatch etc.

File: src/services/CalcFactor.jl [hunk 3]
```diff
@@ -20,44 +20,31 @@ _getDimensionsPartial(fct::DFGFactor) = _getDimensionsPartial(_getCCW(fct))
...
+                    factor = ccwl.usrfnc!,
                     _sampleIdx = 0, 
                     _measCount = length(ccwl.measurement),
                     _legacyMeas = ccwl.measurement, 
                     _legacyParams = ccwl.params, 
                     _allowThreads = true,
-                    cache = ccwl.dummyCache )
+                    cache = ccwl.dummyCache,
+                    fullvariables=ccwl.fullvariables,
+                    solvefor=ccwl.varidx)
```
Comment (dehann) [inline src/services/CalcFactor.jl:36]: we are also going to have to double check this with the multihypo recipes.  Basically the ground rule is that a CalcFactor object represents one specific hypothesis.  If there are particles with different hypos, then each particle gets a different CalcFactor object.

From that, `.fullvariables` and `solvefor` should only refer to the one active hypothesis this CalcFactor object is working with.

File: src/services/CalcFactor.jl [hunk 4]
```diff
@@ -240,11 +224,11 @@ function CommonConvWrapper( usrfnc::T,
                             res::AbstractVector{<:Real}=zeros(zDim),
                             threadmodel::Type{<:_AbstractThreadModel}=MultiThreaded,
                             inflation::Real=3.0,
-                            vartypes=typeof.(getVariableType.(factormetadata.fullvariables)),
+                            vartypes=DataType[],
```
Comment (dehann) [inline src/services/CalcFactor.jl:227]: Why not just:
```suggestion
                            vartypes=typeof.(getVariableType.(fullvariables)),
```

File: src/services/CalcFactor.jl [hunk 5]
```diff
@@ -240,11 +224,11 @@ function CommonConvWrapper( usrfnc::T,
...
                             gradients=nothing,
-                            userCache::CT = nothing ) where {T<:AbstractFactor,P,H,CT}
+                            userCache::CT = nothing,
+                            fullvariables=DFGVariable[] ) where {T<:AbstractFactor,P,H,CT}
```
Comment (dehann) [inline src/services/CalcFactor.jl:230]:
```suggestion
                           ) where {T<:AbstractFactor,P,H,CT}
```

File: src/services/CalcFactor.jl (hunk 3 above)
Comment (dehann) [reply src/services/CalcFactor.jl:36]: > should fullvariables be filtered by hypo

Yes it should be filtered by hypo

File: test/testMultihypoFMD.jl [hunk 6]
```diff
@@ -16,14 +16,14 @@ end
...
-  @warn "getSample(cf::CalcFactor{<:MyFactor},::Int) does not get hypo sub-selected FMD data"
-  @show DFG.getLabel.(cf.metadata.fullvariables)
+  @warn "getSample(cf::CalcFactor{<:MyFactor},::Int) does not get hypo sub-selected FMD data: $(DFG.getLabel.(cf.fullvariables))" cf.solvefor maxlog=1
   # @assert DFG.getLabel.(fmd_[1].fullvariables) |> length < 3 "this factor is only between two variables"
   return rand(cf.factor.Z, 1)
 end
 
 function (cf::CalcFactor{<:MyFactor})(z, X1, X2)
-  @assert DFG.getLabel.(cf.metadata.fullvariables) |> length < 3 "this factor is only between two variables"
+  length(cf.fullvariables) < 3 && @warn("this factor is only between two variables. solvefor=$(cf.solvefor)", maxlog=1)
```
Comment (dehann) [inline test/testMultihypoFMD.jl:25]:
```suggestion
  @assert length(cf.fullvariables) < 3 "this factor is only between two variables. solvefor=$(cf.solvefor)"
```

This error is important in testing

File: src/entities/FactorOperationalMemory.jl (hunk 1 above)
Comment (Affie) [reply src/entities/FactorOperationalMemory.jl:45]: Yes, currently it’s dispatched dynamically in any case and this way it at least saves on compile time.
But the plan is to figure out the best way after the cleanup and not waste time now.

File: src/services/CalcFactor.jl (hunk 3 above)
Comment (Affie) [reply src/services/CalcFactor.jl:36]: Ok. I figured that much, but couldn’t quite see how it was done / can be done without first removing CPT.

File: src/services/CalcFactor.jl (hunk 4 above)
Comment (Affie) [reply src/services/CalcFactor.jl:227]: julia> typeof.(getVariableType.(DFGVariable[]))
Any[]

Comment (Affie) [reply src/services/CalcFactor.jl:227]: But it should be done better

File: test/testMultihypoFMD.jl (hunk 6 above)
Comment (Affie) [reply test/testMultihypoFMD.jl]: Jip, I fixed it, was just to test with and help me understand.

File: src/services/CalcFactor.jl (hunk 4 above)
Comment (Affie) [reply src/services/CalcFactor.jl]: done

**Comment by Affie** [comment]
We should probably tag a new patch release before merging this into master.

**Pull request merged**